### PR TITLE
[Doc] Add mention to shared section in `config_for` docs [ci skip]

### PR DESCRIPTION
**Pull request by albertoalmagro**
### Summary
#37913 added the possibility to deeply merge configurations by grouping them within a shared section. This powerful alternative was not reflected in any documentation, which made my team think it was not possible until I found out this feature after looking at the source code.

This patch reflects this change in the documentation so that it is easier for other developers to know about this behavior.

### CODE REVIEW

File: guides/source/configuring.md [hunk 1]
```diff
@@ -1617,6 +1617,27 @@ You can also use `Rails::Application.config_for` to load whole configuration fil
   Rails.configuration.payment['merchant_id'] # => production_merchant_id or development_merchant_id
   ```
 
+`Rails::Application.config_for` can also merge configurations deeply. To achieve
```
Comment (eileencodes) [inline guides/source/configuring.md:1620]: I think we need to reword "merge configurations deeply". If I was a beginner I'd have no idea what this means. 

I'd reword this to: "`Rails::Application.config_for` supports a `shared` configuration to group common configurations. The shared configuration will be merged into the environment configuration." or something similar.

File: guides/source/configuring.md [hunk 2]
```diff
@@ -1617,6 +1617,27 @@ You can also use `Rails::Application.config_for` to load whole configuration fil
...
+that behaviour simply group common configurations within a shared section:
+
+  ```yaml
+  # config/example.yml
+  shared:
+    foo:
+      bar:
+        baz: 1
```
Comment (eileencodes) [inline guides/source/configuring.md:1628]:
```suggestion
        baz: 1
        
```

File: railties/lib/rails/application.rb [hunk 3]
```diff
@@ -220,6 +220,22 @@ def message_verifier(verifier_name)
     #     Rails.application.configure do
     #       config.middleware.use ExceptionNotifier, config_for(:exception_notification)
     #     end
+    #
+    #     # It also merges configurations in a shared section deeply:
```
Comment (eileencodes) [inline railties/lib/rails/application.rb:224]: The same change from above should be made here. We should only say "You can also store configurations in a shared section which will be merged with the environment configuration".

File: railties/lib/rails/application.rb [hunk 4]
```diff
@@ -220,6 +220,22 @@ def message_verifier(verifier_name)
...
+    #
+    #     # config/example.yml
+    #     shared:
+    #       foo:
+    #         bar:
+    #           baz: 1
```
Comment (eileencodes) [inline railties/lib/rails/application.rb:230]:
```suggestion
    #           baz: 1
    
```

**Review by eileencodes — CHANGES_REQUESTED** [review]

**Review by albertoalmagro — COMMENTED** [review]
Thanks @eileencodes for your review! As a non-native English speaker I am more than happy to accept all your suggestions.

**Review by eileencodes — APPROVED** [review]

**Comment by eileencodes** [comment]
Thanks @albertoalmagro!
